File: hs/src/Day1.hs
module Day1 (day1) where

import Data.Text (Text)
import Data.Text qualified as T

day1 :: Text -> (Int, Int)
day1 input = (zeros, zeros + turns)
  where
    (_, zeros, turns) = foldl' go (50, 0, 0) $ T.lines input
    go k ('L' T.:< line) = go' k $ -read (T.unpack line)
    go k ('R' T.:< line) = go' k $ read (T.unpack line)
    go k _ = k
    go' (pos, zeros, turns) rotation =
      ( pos'',
        (if pos'' == 0 then succ else id) zeros,
        (if pos /= 0 && pos' < 0 then succ else id) turns'
      )
      where
        pos' = pos + rotation
        pos'' = pos' `mod` 100
        turns' = turns + abs (pos' - signum pos') `div` 100

File: rs/src/day1.rs
pub fn solve(input: &str) -> (u32, u32) {
    let (mut pos, mut zeros, mut turns) = (50, 0, 0);
    for line in input.lines() {
        let rotation = if let Some(line) = line.strip_prefix('L') {
            -line.parse::<i32>().unwrap()
        } else if let Some(line) = line.strip_prefix('R') {
            line.parse().unwrap()
        } else {
            continue;
        };
        let pos2 = pos + rotation;
        turns += pos2.abs_diff(rotation.signum()) / 100 + (pos != 0 && pos2 < 0) as u32;
        pos = pos2.rem_euclid(100);
        zeros += (pos == 0) as u32;
    }
    (zeros, zeros + turns)
}
